jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ServerShutdown.java

```diff
@@ -0,0 +1,518 @@
+            Boolean.parseBoolean(System.getProperty("STOP.EXIT", "true")));
+    }
+
+    /**
+     * Create a new ShutdownMonitor.
+     *
+     * @param host the host
+     * @param port the port
+     * @param key the key that must be passed to allow a shutdown
+     * @param exitVm flag to exit vm on successful shutdown
+     */
+    public ServerShutdown(@Name("host") String host,
+                          @Name("port") int port,
+                          @Name("key") String key,
+                          @Name("exitVm") boolean exitVm)
+    {
+        this.host = host;
+        this.port = port;
+        this.key = key;
+        this.exitVm = exitVm;
+    }
+
+    @Override
+    public void addComponent(LifeCycle component)
+    {
+        try (AutoLock l = lock.lock())
+        {
+            // TODO: should we reject calls to this method after the ServerSocket is started?
+            this.components.add(component);
+        }
+    }
+
+    @Override
+    public boolean removeComponent(LifeCycle component)
+    {
+        // TODO: should we reject calls to this method after the ServerSocket is started?
+        boolean ret = false;
+        try (AutoLock l = lock.lock())
+        {
+            ret = this.components.remove(component);
+        }
+
+        return ret;
+    }
+
+    public boolean hasComponent(LifeCycle component)
+    {
+        boolean ret = false;
+        try (AutoLock l = lock.lock())
+        {
+            ret = this.components.contains(component);
+        }
+
+        return ret;
+    }
+
+    @Override
+    public void start() throws Exception
+    {
+        try (AutoLock l = lock.lock())
+        {
+            if (serverSocket != null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Already started: {}", this);
+                return; // cannot start it again
+            }
+
+            listen();
+            if (serverSocket != null)
+            {
+                Thread thread = new Thread(new ServerShutdown.ShutdownRunnable(serverSocket));
+                thread.setDaemon(true);
+                thread.setName(TypeUtil.toShortName(this.getClass()));
+                thread.start();
+            }
+        }
+    }
+
+    @Override
+    public void stop()
+    {
+        IO.close(serverSocket);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Closed ServerSocket");
+        serverSocket = null;
+    }
+
+    public boolean isExitVm()
+    {
+        return exitVm;
+    }
+
+    public String getHost()
+    {
+        return host;
+    }
+
+    public String getKey()
+    {
+        return key;
+    }
+
+    public void setPort(int port)
+    {
+        if (port < 0 || port > 0xFFFF)
+        {
+            throw new IllegalArgumentException("Invalid Port: " + port);
+        }
+        this.port = port;
+    }
+
+    public int getPort()
+    {
+        return port;
+    }
+
+    public boolean isConfigurationValid()
```

Comment:
If this really is a check if the configuration is valid, then it should be done in `start()` method, never called externally. However, I think what it _really_ is, is checking whether or not the shutdown service has been _enabled_, indicated by configuring of the equivalent to the current `STOP.PORT` system property. So this method is better named `isEnabled()`.